apps/designer/app/shared/db/build.server.ts

```diff
@@ -0,0 +1,34 @@
+import { prisma, Build as DbBuild } from "@webstudio-is/prisma-client";
+import { z } from "zod";
+
+const Page = z.object({
+  id: z.string(),
+  name: z.string(),
+  path: z.string(),
+  title: z.string(),
+  meta: z.record(z.string(), z.string()),
+});
+const Pages = z.array(Page);
+type Pages = z.infer<typeof Pages>;
```

Comment:
In Webflow they have Folder as separate type from Page. It has no content, and only `name` and `slug` as properties. Should we do the same? Or do we want folders to also act as pages?

As a user it feels kinda like I'm forced to create some content for pages that I don't need. For example, if I'm building a blog I might want following pages:

 - `/` — list of posts
 - `/posts/first-post`
 - `/posts/second-post`

And I don't want anything at `/posts/`

Reply:
Might be a good idea to have `Homepage` separate from other pages:

```ts
type Pages = {
  homepage: Page;
  pages: array<Folder | Page>;
}
```

This will enforce 2 requirements:

 - Homepage must exist
 - Homepage must not be nested in any folder

Reply:
good idea

Reply:
> In Webflow they have Folder as separate type from Page. It has no content, and only name and slug as properties. Should we do the same? Or do we want folders to also act as pages?


@taylornowotny do you think this would be useful? I was also thinking about it and couldn't think of a use case

Reply:
> As a user it feels kinda like I'm forced to create some content for pages that I don't need. For example, if I'm building a blog I might want following pages:
> 
> / — list of posts
> /posts/first-post
> /posts/second-post
> And I don't want anything at /posts/

We failed to mention the "path" we let user enter is not a slug, user can enter a full path, folder doesn't create a path in our case, that's why having folders in our case is actually like folders: If you want a list you can still use /posts in the posts page

Reply:
One thing that is a little weird is that you would have a page called "posts" on the same level as the pages for each specific post.

Reply:
> We failed to mention the "path" we let user enter is not a slug, user can enter a full path

Hm, why do we need folders at all then? What if we implement pages as a flat list, at least for the first release? Would safe us a lot of time.

Reply:
Although if we later decide to control pages' URLs via nesting in folders, it will be hard to migrate to that from arbitrary paths.

Reply:
I have been back and forth on this too. Folders adding to the path force a certain URL structure you might not want to have, this is the feedback I got from @taylornowotny .

At the same time folders are good for organizing pages into categories of some sort when you have a lot of them. 

I was wondering if we could go with flat list of pages for beta too and wasn't sure if its going to save us a lot of time if we go with the same datastructure in the database.

The thing is if we want to leverage the patches, we would still need to update a single object. Potentially we could be updating one page at a time, but never the entire pages object. Then we have to ensure the right sort order. Having a single data structure for all pages ensures sort order automatically.

If we go with a flat list now, how hard will be adding folders in the future?

Thoughts?

Reply:
I don't think we truly need folders for beta release, but I was thinking its not gonna save us much time if we don't have them

Reply:
Regarding folder enforcing a path: I guess we could even make it an option. By default folder would have no path, but if added, it gets added automatically to all pages inside that folder, wdyt @taylornowotny ?

Reply:
Another thought: what if pages and folders are the same? If user doesn't want to use folder as a page, they will simply not use it. Nothing forces them to use that page, they are linking the pages anyways manually. So if its not linked its not used ... ? @taylornowotny

Reply:
> The thing is if we want to leverage the patches, we would still need to update a single object. Potentially we could be updating one page at a time, but never the entire pages object. Then we have to ensure the right sort order. Having a single data structure for all pages ensures sort order automatically.

Yeah, I'd still kept them in one JSON object event if they a flat list.

> If we go with a flat list now, how hard will be adding folders in the future?

A bit hard to tell, but I feel like if folders are used only for organizing, then there won't be much additional work compared to doing it now: `work to implement now with folders ~= now without folders + folders later`.

And this will allow us to think more about this. Maybe we'll decide to have tags instead of folders, or something like that.

Reply:
yeah, the question is if it saves us much time, in that case, but yeah, thinking through all those questions properly would be highly benefitial

Reply:
>  the question is if it saves us much time

Hard to tell. I think from a day to a week

Reply:
Ok lets first figure out without pages and then see if we can bring them in within a few days right away. We would still use the same tree component for drag&drop etc, so supporting dragging between folders should be easy

Reply:
Re: folders determining the path - I don't think is something we need to include as an option. Maybe some users would want to use folders to organize paths (like "/folder/page") but since they have full control over the path on every page anyway, they can set that up manually.

Reply:
Folders are meant to be purely organizational

Reply:
We could use tags instead of folders, but I think tags would be an inferior organizational tool in this case for the following reasons:

- folders are better for visual organization in the pages list because of nesting. Users can nest pages to hide them from view while they focus on other pages. With tags there is no nesting
- both tags and folders are good for grouping pages, but there's not currently a UI for tags so that would have to be designed and built
- tags are best used with sorting and filtering controls, which are best used for large sets. An average website might have a dozen pages or less, so the full utility of tags is wasted on such a small list.